django/test/runner.py

```diff
@@ -856,17 +928,64 @@ def find_top_level(top_level):
     return top_level
 
 
-def reorder_tests(tests, classes, reverse=False):
+def _class_shuffle_key(cls):
+    return f'{cls.__module__}.{cls.__qualname__}'
+
+
+def shuffle_tests(tests, shuffler):
     """
-    Reorder an iterable of tests by test type, removing any duplicates.
+    Return an iterator over the given tests in a shuffled order, keeping tests
+    next to other tests of their class.
+
+    `tests` should be an iterable of tests.
+    """
+    tests_by_type = {}
+    for _, class_tests in itertools.groupby(tests, type):
+        class_tests = list(class_tests)
+        test_type = type(class_tests[0])
+        class_tests = shuffler.shuffle(class_tests, key=lambda test: test.id())
+        tests_by_type[test_type] = class_tests
+
+    classes = shuffler.shuffle(tests_by_type, key=_class_shuffle_key)
+
+    return itertools.chain(*(tests_by_type[cls] for cls in classes))
+
+
+def reorder_test_bin(tests, shuffler=None, reverse=False):
+    """
+    Return an iterator that reorders the given tests, keeping tests next to
+    other tests of their class.
+
+    `tests` should be an iterable of tests that supports reversed().
+    """
+    if shuffler is None:
+        if reverse:
+            return reversed(tests)
+        # The function must return an iterator.
+        return iter(tests)
+
+    tests = shuffle_tests(tests, shuffler)
+    if not reverse:
+        return tests
+    # Arguments to reversed() must be reversible.
+    return reversed(list(tests))
 
-    `classes` is a sequence of types. The result is returned as an iterator.
 
+def reorder_tests(tests, classes, reverse=False, shuffler=None):
+    """
+    Reorder an iterable of tests by test type, removing any duplicates.
+
+    The result is returned as an iterator. `classes` is a sequence of types.
     All tests of type classes[0] are placed first, then tests of type
     classes[1], etc. Tests with no match in classes are placed last.
 
     If `reverse` is True, sort tests within classes in opposite order but
     don't reverse test classes.
+
+    The `shuffler` argument is an optional instance of this module's `Shuffler`
+    class.  If provided, tests will be shuffled within each `classes` group,
+    but keeping tests with other tests of their TestCase class.  Reversing is
```

Comment:
Remove extra whitespaces:
```suggestion
    class. If provided, tests will be shuffled within each `classes` group,
    but keeping tests with other tests of their TestCase class. Reversing is
```